mm2src/coins/lp_coins.rs

```diff
@@ -3604,32 +3645,198 @@ impl MmCoinStruct {
+    /// No dex fee is taken (if taker is dex pubkey)
+    NoFee,
     /// Standard dex fee which will be sent to the dex fee address
     Standard(MmNumber),
-    /// Dex fee with the burn amount.
-    ///   - `fee_amount` goes to the dex fee address.
-    ///   - `burn_amount` will be added as `OP_RETURN` output in the dex fee transaction.
+    /// Dex fee with the burn amount
     WithBurn {
+        /// Amount to go to the dex fee address
         fee_amount: MmNumber,
+        /// Amount to be burned
         burn_amount: MmNumber,
+        /// This indicates how to burn the burn_amount
+        burn_destination: DexFeeBurnDestination,
     },
 }
 
+/// Option to enforce Standard or WithBurn creation
+pub enum DexFeeOption {
+    Standard,
+    WithBurn,
+}
+
 impl DexFee {
-    /// Creates a new `DexFee` with burn amounts.
-    pub fn with_burn(fee_amount: MmNumber, burn_amount: MmNumber) -> DexFee {
-        DexFee::WithBurn {
-            fee_amount,
-            burn_amount,
+    const DEX_FEE_SHARE: &str = "0.75";
+
+    /// Recreates a `DexFee` from separate fields (usually stored in db).
+    #[cfg(any(test, feature = "for-tests"))]
+    pub fn create_from_fields(fee_amount: MmNumber, burn_amount: MmNumber, ticker: &str) -> DexFee {
+        if fee_amount == MmNumber::default() && burn_amount == MmNumber::default() {
+            return DexFee::NoFee;
+        }
+        if burn_amount > MmNumber::default() {
+            let burn_destination = match ticker {
+                "KMD" => DexFeeBurnDestination::KmdOpReturn,
+                _ => DexFeeBurnDestination::PreBurnAccount,
+            };
+            DexFee::WithBurn {
+                fee_amount,
+                burn_amount,
+                burn_destination,
+            }
+        } else {
+            DexFee::Standard(fee_amount)
+        }
+    }
+
+    /// Creates a new `DexFee` for a taker coin to sell.
+    fn new_from_taker_coin(
+        taker_coin: &dyn MmCoin,
+        rel_ticker: &str,
+        trade_amount: &MmNumber,
+        taker_pubkey: Option<&[u8]>,
+        opt: DexFeeOption,
+    ) -> DexFee {
+        if let Some(taker_pubkey) = taker_pubkey {
+            if !taker_coin.is_privacy() && taker_coin.dex_pubkey() == taker_pubkey {
+                // Only updated taker nodes will use this feature (no need to activate by version)
+                return DexFee::NoFee; // no dex fee if the taker is the dex pubkey
+            }
+        }
+        // calc dex fee
+        let rate = Self::dex_fee_rate(taker_coin.ticker(), rel_ticker);
+        let dex_fee = trade_amount * &rate;
+        let min_tx_amount = MmNumber::from(taker_coin.min_tx_amount());
+        if dex_fee <= min_tx_amount {
+            return DexFee::Standard(min_tx_amount);
+        }
+
+        if taker_coin.is_kmd() {
+            // use a special dex fee option for kmd
+            let (fee_amount, burn_amount) = Self::calc_burn_amount_for_op_return(&dex_fee, &min_tx_amount);
```

Comment:
> if we came to find the burn_amount = 0 in this branch we should return dex fee standard

This changes already working code. I guess we could do this only as version activated feature

Reply:
> tbh i think we should merge Standard & WithBurn to something general

Eventually we will eliminate the Standard option (when all nodes are upgraded), so in fact this will be done.
For now I suggest to retain it as it would be easier to eliminated the code related to Standard.